// wx-server-sdk/index.js
/**
 * Mock wx-server-sdk for smoke tests
 * 提供最小化 mock 使得云函数可以被 require()
 */
'use strict';

const mockCollectionMethods = () => ({
  where:       () => mockCollectionMethods(),
  field:       () => mockCollectionMethods(),
  limit:       () => ({ get: () => Promise.resolve({ data: [] }) }),
  orderBy:     () => ({ get: () => Promise.resolve({ data: [] }) }),
  get:         () => Promise.resolve({ data: [] }),
  add:         ({ data }) => Promise.resolve({ _id: 'mock_' + Date.now() }),
  count:       () => Promise.resolve({ total: 0 }),
  update:      ({ data }) => Promise.resolve({ stats: { updated: 1 } }),
});

const mockDb = {
  collection:  () => mockCollectionMethods(),
  command: {
    eq:        v => v,
    neq:       v => ({ $neq: v }),
    in:        arr => ({ $in: arr }),
    nin:       arr => ({ $nin: arr }),
    gt:        v => ({ $gt: v }),
    gte:       v => ({ $gte: v }),
    lt:        v => ({ $lt: v }),
    lte:       v => ({ $lte: v }),
    inc:       v => ({ $inc: v }),
    push:      v => ({ $push: v }),
    elemMatch: v => ({ $elemMatch: v }),
    // command.RegExp — 云函数通过 db.command.RegExp() 访问
    RegExp:    ({ regexp, options }) => new RegExp(regexp, options),
  },
  // db.RegExp — 云函数通过 db.RegExp() 直接访问时使用
  RegExp:    ({ regexp, options }) => new RegExp(regexp, options),
  serverDate:  () => new Date(),
};

const mockCloud = {
  openapi: {
    phonenumber: {
      getPhoneNumber: () => Promise.resolve({
        errCode: 0,
        phoneInfo: { purePhoneNumber: 'mock_phone' },
      }),
    },
  },
};

module.exports = {
  init:               () => {},
  database:           () => mockDb,
  DYNAMIC_CURRENT_ENV: 'mock-env',
  getWXContext:       () => ({ OPENID: 'mock-openid', APPID: 'mock-appid' }),
  cloud:              mockCloud,
};
